codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -335,6 +340,236 @@ private void writeSafeFieldAccessor(
+
+    private void generateOperationHttpBindingDeserializer(
+            GenerationContext context,
+            OperationShape operation
+    ) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Model model = context.getModel();
+        GoWriter writer = context.getWriter();
+
+        Shape outputShape = model.expectShape(operation.getOutput()
+                .orElseThrow(() -> new CodegenException(
+                        "missing output shape for operation: " + operation.getId())));
+
+        HttpBindingIndex bindingIndex = model.getKnowledge(HttpBindingIndex.class);
+        Map<String, HttpBinding> bindingMap = bindingIndex.getResponseBindings(operation).entrySet().stream()
+                .filter(entry -> isRestBinding(entry.getValue().getLocation()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> {
+                    throw new CodegenException("found duplicate binding entries for same response operation shape");
+                }, TreeMap::new));
+
+        // do not generate if no HTTPBinding for operation output
+        if (bindingMap.size() == 0) {
+            return;
+        }
+
+        generateShapeDeserializerFunction(writer, model, symbolProvider, outputShape, bindingMap);
+    }
+
+    private void generateErrorHttpBindingDeserializer(
+            GenerationContext context,
+            ShapeId errorBinding
+    ) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Model model = context.getModel();
+        GoWriter writer = context.getWriter();
+        HttpBindingIndex bindingIndex = model.getKnowledge(HttpBindingIndex.class);
+        Shape errorBindingShape = model.expectShape(errorBinding);
+
+        Map<String, HttpBinding> bindingMap = bindingIndex.getResponseBindings(errorBinding).entrySet().stream()
+                .filter(entry -> isRestBinding(entry.getValue().getLocation()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> {
+                    throw new CodegenException("found duplicate binding entries for same error shape");
+                }, TreeMap::new));
+
+        // do not generate if no HTTPBinding for Error Binding
+        if (bindingMap.size() == 0) {
+            return;
+        }
+
+        generateShapeDeserializerFunction(writer, model, symbolProvider, errorBindingShape, bindingMap);
+    }
+
+
+    private void generateShapeDeserializerFunction(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            Shape targetShape,
+            Map<String, HttpBinding> bindingMap
+    ) {
+        Symbol targetSymbol = symbolProvider.toSymbol(targetShape);
+        Symbol smithyHttpResponsePointableSymbol = SymbolUtils.createPointableSymbolBuilder(
+                "Response", GoDependency.SMITHY_HTTP_TRANSPORT).build();
+
+        writer.addUseImports(smithyHttpResponsePointableSymbol);
```

Comment:
This should not be necessary - it should be automatic when you write out the symbol.

Reply:
Right, using $P so that should import directly.